Cxx examples

darbyjohnston opened the pull request:
```Fixes #917```

This is WIP code to convert the Python examples to C++. I've successfully converted "summarize_timing.py" and started converting "flatten_video_tracks.py". With these changes, and the CMake variable OTIO_CXX_EXAMPLE set to "ON", you can run:

> ./examples/summarize_timing ../tests/sample_data/premiere_example.xml
> ./examples/flatten_video_tracks ../tests/sample_data/premiere_example.xml tmp.xml

Converting the examples brought up a couple of questions:
* It looks like the Python API has diverged a bit from C++? For example what are the C++ equivalents of each_clip() and deepcopy()?
* How should a user of the C++ API handle memory management? The "C++ Implementation Details" documentation mentions the use of possibly_delete() and Retainers?
* Can a user of the C++ API utilize the Python adapters for file I/O? I've implemented a utility class that provides this, and it seems to work OK initially, but is it the right approach? If so maybe the code should be moved into the OTIO library for general usage?

Note this code currently only works for Linux/macOS, there is some additional code that needs to be added for Windows.

darbyjohnston commented:
I've updated the code to use Retainers and added the missing code for Windows.

meshula commented on examples/summarize_timing.cpp:
```diff
@@ -0,0 +1,131 @@
+{
+    for (auto effect : item.value->effects())
+    {
+        if (auto freezeFrame = dynamic_cast<otio::FreezeFrame*>(effect.value))
```
per your notes elsewhere, this definitely begs a dynamic_retained_cast.

darbyjohnston replied:
Maybe something like:
```
template <class T, class U>
otio::SerializableObject::Retainer<T> dynamic_retainer_cast(otio::SerializableObject::Retainer<U> const& retainer)
{
    return dynamic_cast<T*>(retainer.value);
}
```

meshula replied:
yes, exactly that ~ I think I'm leaning to sending this over to OTIO as a PR. It's in some sense trivial, but it would be nice to do it the same everywhere

darbyjohnston replied:
It might also be nice to add the "->" operator to Retainer so things like this:
`timeline.value->tracks()`

Could be simplified to:
`timeline->tracks()`

meshula replied:
I like it

darbyjohnston replied:
@meshula Just double-checking, did you want me to open a PR for the dynamic_retainer_cast()? If so do you think the other variants like static_retainer_cast(), etc., are useful as well?

ssteinbach replied:
Just so I understand, that would make this code:
```c++
if (auto freezeFrame = otio::dynamic_retainer_cast<otio::FreezeFrame*>(effect)) {
```

darbyjohnston replied:
Actually even a little bit simpler since the "*" isn't necessary:
```
if (auto freezeFrame = otio::dynamic_retainer_cast<otio::FreezeFrame>(effect)) {
```

darbyjohnston replied:
I just created a draft PR that add this:
https://github.com/PixarAnimationStudios/OpenTimelineIO/pull/923

meshula commented on examples/flatten_video_tracks.cpp:
```diff
@@ -0,0 +1,74 @@
+    }
+    
+    // Read the file
+    PythonAdapters adapters;
```
This is a cool thing, but I'd suggest breaking out a separated "embedding a python interpreter" example, rather than using it as a utility in cpp in general. I can imagine people starting to proliferate python embeddings via copy-pasta, not understanding whether that's a requirement of OTIO or not. Reading and writing a JSON file via C++ is supported out of the box, so in terms of being instructive, the latter is the approach I'd recommend.

jminor replied:
In the past we've encouraged people to execute `otioconvert` or `otiocat` as sub-processes to avoid the subject of embedding Python in their application. Example programs that show how to accomplish either or both of these would be very helpful. Then folks can pick whichever suits their use case.

jminor replied:
Also, using `otiocat` (or equivalent code via embedded Python) would avoid the need for a temp file when reading. You could probably avoid the temp file when writing also.

darbyjohnston replied:
Sure thing, I'll make a dedicated example like you suggest and the other examples will just work with JSON files.

darbyjohnston replied:
Thanks Josh, I'll checkout otioconvert/otiocat, I wasn't thrilled with the temp file approach.

meshula commented on examples/summarize_timing.cpp:
```diff
@@ -0,0 +1,131 @@
+{
+    if (errorStatus.outcome != otio::ErrorStatus::Outcome::OK)
+    {
+        std::cout << '\t' << label << ": None" << std::endl;
```
std::cerr would be appropriate here (and wherever errors are reported)

meshula commented on examples/summarize_timing.cpp:
```diff
@@ -0,0 +1,131 @@
+
+int main(int argc, char** argv)
+{
+    PythonAdapters adapters;
```
same comment, on using native facilities here

meshula reviewed (commented):
This looks very helpful, I've dropped a few notes inline.

meshula commented:
Might be an idea to mark this PR as "WIP" until it's ready to land? That's the convention we're using in this repo, in general, when things are being iterated and thought through.

ssteinbach commented:
Rather than the `wip` tag, marking it as a draft PR is even better.  Then we get a nice notification when its no longer draft.

darbyjohnston commented:
Any thoughts about the Python functions each_clip() and deepcopy() for C++?

meshula commented:
each_clip() and deepcopy() both seem handy for c++ use. Maybe a quick sketch and discussion if they should be a topic for a PR for OTIO? (Since deepcopy() is a Pythonism, I'd suggest deep_copy for C++.)

darbyjohnston commented:
I've broken out the C++ code for using Python adapters into two separate examples, "python_adapters_child_process.cpp" and "python_adapters_embed.cpp", for using the adapters via child processes and an embedded Python interpreter respectively.

I think I prefer the embedding approach, that way there is less environment setup to worry about (making sure otioconvert is in the search path), and there is less platform specific code (child process creation and pipes seem more complicated on Windows).

I still need to take a look and see if I can get rid of the temporary files, and the flatten_video_tracks.cpp example is missing a C++ version of deepcopy().

@meshula Have you used CreateProcess() on Windows before? The MSDN documentation says to pass "cmd.exe" in to execute bat files, but I got the error "The system cannot find the file specified.". If I use "C:\windows\system32\cmd.exe" it works OK, but it seems like a bad idea to hard code the path...

meshula commented:
The embedding approach is valid, but it's a whole other can of worms. Not everyone will want to bind Python into their application, because there's so much other environment to set up to make that fully functional. Embedded IS interesting, but so is calling out. 

CreateProcess with a full path to system32\cmd.exe is actually fine; cmd has been there forever, and is unlikely to move in the foreseeable future. It's like the assumption that we make in expecting to be able to find sh at /bin/sh until the end of time.

darbyjohnston commented:
OK, I've changed the example "python_adapters_embed.cpp" to use Python directly without needing temporary files. I did this by converting the timeline into JSON and passing that to/from Python where the adapters are used for file IO. It's a bit more code than before, but it's not too bad and doesn't require any platform specific code.

I also looked into deepcopy() and apparently it's just a wrapper around SerializableObject::clone(), so maybe it's not needed for C++. I updated the example "flatten_video_tracks.cpp" to use clone() and it should be complete now.

I think that addresses all of the feedback, should I "un-draft" this PR? Or should this also include C++ versions of the rest of the Python examples?

ssteinbach commented on CMakeLists.txt:
```diff
@@ -153,6 +153,8 @@ if(WIN32)
     set(OTIO_DEBUG_POSTFIX "d")
 endif()
 
+set_property(GLOBAL PROPERTY USE_FOLDERS ON)
```
just googled what this does, neat.

darbyjohnston replied:
It definitely helps keep things neat in IDEs like Visual Studio, though one downside is you need to manually specify the folders for each target:
`set_target_properties(${example} PROPERTIES FOLDER examples)`

ssteinbach commented on CMakeLists.txt:
```diff
@@ -207,3 +209,8 @@ add_subdirectory(src/opentimelineio)
     add_subdirectory(src/py-opentimelineio)
 endif()
+
+if(OTIO_CXX_EXAMPLES AND OTIO_PYTHON_INSTALL)
+    add_subdirectory(examples)
+endif()
```
Should this write an error if `OTIO_CXX_EXAMPLES` is on but `OTIO_PYTHON_INSTALL` is off?  Or a message of some kind?

darbyjohnston replied:
Actually now that the C++/Python code is in separate examples, maybe this should be changed to just:
```
if(OTIO_CXX_EXAMPLES)
    add_subdirectory(examples)
endif()
```
And then the examples/CMakeLists.txt could optionally add the C++/Python code when Python is enabled:
```
if(OTIO_PYTHON_INSTALL)
    list(APPEND examples python_adapters_child_process)
    list(APPEND examples python_adapters_embed)
endif()
```
That way the pure C++ examples can be built without Python.

ssteinbach replied:
Ah yeah -- lets do that.  That sounds simpler and clearer.  Thanks!

ssteinbach reviewed (commented):
Thanks for putting this together, @darbyjohnston.  Both `deepcopy` and `each_clip` are based on pythonisms, and yeah -- deep copy() is just exposing the `clone` functionality.  I wonder if `each_clip` would become a C++11 style iterator?  Or are those just not good?

meshula commented:
The interface for a C++ style iterator is not hard to satisfy; here's a nice overview. https://internalpointers.com/post/writing-custom-iterators-modern-cpp

That gives you compatibility with lots of STL features like std::copy, so iterators would be a definite nice-to-have.

darbyjohnston commented:
Un-drafting this since it seems like most of the feedback has been resolved and the other work for Retainers and each_clip() should probably be separate PRs...

ssteinbach reviewed (approved):
Looks good to me!  Thanks @darbyjohnston!
